File: nn/src/pgn.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Iterator


@dataclass
class ParsedGame:
    headers: dict[str, str]
    movetext_raw: str


_UCI_TOKEN = re.compile(r"^[a-i][0-9][a-i][0-9][a-z]?$")
_ICCS_MOVE = re.compile(
    r"\b([A-Ia-i]\d+)\s*-\s*([A-Ia-i]\d+)\b",
    re.IGNORECASE,
)
_HEADER_RE = re.compile(r'^\[(\w+)\s+"((?:\\.|[^"])*)"\s*\]$')


def strip_comments_and_variations(movetext: str) -> str:
    """去掉 { } 注释与 ( ) 变着，保留主干着法串。"""
    s = movetext
    while "{" in s:
        s = re.sub(r"\{[^{}]*\}", " ", s)
    depth = 0
    out: list[str] = []
    for ch in s:
        if ch == "(":
            depth += 1
            continue
        if ch == ")":
            depth = max(0, depth - 1)
            continue
        if depth == 0:
            out.append(ch)
    return "".join(out)


def movetext_uci_tokens(movetext: str) -> list[str]:
    """从清理后的 movetext 中提取 **pyffish 风格** 四格 UCI（极少见于本仓库 PGN，保留兼容）。"""
    clean = strip_comments_and_variations(movetext)
    tokens: list[str] = []
    for raw in clean.split():
        t = raw.strip()
        if not t:
            continue
        if t in ("1-0", "0-1", "1/2-1/2", "*"):
            continue
        if re.match(r"^\d+\.\.\.", t):
            continue
        if re.match(r"^\d+\.", t):
            continue
        if _UCI_TOKEN.match(t):
            tokens.append(t)
    return tokens


def movetext_iccs_pairs(movetext: str) -> list[str]:
    """
    提取 ICCS 着法 `C3-C4`（东萍 / 联赛 PGN 常见）。
    返回小写半格、带连字符，如 `c3-c4`，便于再转为引擎 UCI 或 pyffish。
    """
    clean = strip_comments_and_variations(movetext)
    out: list[str] = []
    for m in _ICCS_MOVE.finditer(clean):
        a, b = m.group(1), m.group(2)
        out.append(f"{a.lower()}-{b.lower()}")
    return out


def pgn_format(game_headers: dict[str, str]) -> str:
    """`ICCS` | `WXF` | `UCI` | 空（空则按棋谱推断）。"""
    return (game_headers.get("Format") or "").strip().upper()


def read_pgn_games(path: str | Path) -> Iterator[ParsedGame]:
    """
    读取 PGN：以 `[Event` 作为新对局起点；每个对局内标签头读到首个空行，
    其后直至下一 `[Event`（或文件尾）为 movetext（可与头块之间隔一个空行）。
    """
    raw = Path(path).read_text(encoding="utf-8", errors="replace")
    raw = raw.replace("\r\n", "\n").replace("\r", "\n")
    starts = [m.start() for m in re.finditer(r"(?m)^\[Event\s", raw)]
    if not starts:
        return
    starts.append(len(raw))
    for i in range(len(starts) - 1):
        chunk = raw[starts[i] : starts[i + 1]].strip()
        if not chunk:
            continue
        lines = chunk.split("\n")
        headers: dict[str, str] = {}
        movetext_lines: list[str] = []
        phase = "headers"
        for line in lines:
            s = line.strip()
            if phase == "headers":
                if not s:
                    phase = "moves"
                    continue
                if s.startswith("["):
                    m = _HEADER_RE.match(s)
                    if m:
                        key, val = m.group(1), m.group(2).replace('\\"', '"')
                        headers[key] = val
                    continue
                phase = "moves"
            if phase == "moves" and s:
                movetext_lines.append(s)
        if headers:
            yield ParsedGame(headers=headers, movetext_raw=" ".join(movetext_lines).strip())

File: crates/xiangqi_dataset/src/pgn.rs
//! PGN 读取与 movetext 清理（对齐 `nn/src/pgn.py`）。

use regex::Regex;
use std::sync::OnceLock;

#[derive(Debug, Clone)]
pub struct ParsedGame {
    pub headers: std::collections::HashMap<String, String>,
    pub movetext_raw: String,
}

fn header_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r#"^\[(\w+)\s+"((?:\\.|[^"])*)"\s*\]$"#).expect("header re"))
}

fn iccs_move_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"\b([A-Ia-i]\d+)\s*-\s*([A-Ia-i]\d+)\b").expect("iccs re"))
}

fn uci_token_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"^[a-i][0-9][a-i][0-9][a-z]?$").expect("uci re"))
}

fn move_num_dot_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"^\d+\.").expect("move num"))
}

fn black_move_num_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"^\d+\.\.\.").expect("black move num"))
}

fn brace_comment_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"\{[^{}]*\}").expect("brace"))
}

fn event_line_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"(?m)^\[Event\s").expect("event"))
}

/// 去掉 `{ }` 与 `( )` 变着。
pub fn strip_comments_and_variations(movetext: &str) -> String {
    let mut s = movetext.to_string();
    while s.contains('{') {
        s = brace_comment_re().replace_all(&s, " ").to_string();
    }
    let mut depth = 0i32;
    let mut out = String::new();
    for ch in s.chars() {
        match ch {
            '(' => depth += 1,
            ')' => depth = (depth - 1).max(0),
            _ => {
                if depth == 0 {
                    out.push(ch);
                }
            }
        }
    }
    out
}

pub fn movetext_uci_tokens(movetext: &str) -> Vec<String> {
    let clean = strip_comments_and_variations(movetext);
    let mut tokens = Vec::new();
    let uci_re = uci_token_re();
    let md = move_num_dot_re();
    let bd = black_move_num_re();
    for raw in clean.split_whitespace() {
        let t = raw.trim();
        if t.is_empty() {
            continue;
        }
        if matches!(t, "1-0" | "0-1" | "1/2-1/2" | "*") {
            continue;
        }
        if bd.is_match(t) || md.is_match(t) {
            continue;
        }
        if uci_re.is_match(t) {
            tokens.push(t.to_string());
        }
    }
    tokens
}

pub fn movetext_iccs_pairs(movetext: &str) -> Vec<String> {
    let clean = strip_comments_and_variations(movetext);
    let mut out = Vec::new();
    for m in iccs_move_re().captures_iter(&clean) {
        let a = m.get(1).unwrap().as_str().to_lowercase();
        let b = m.get(2).unwrap().as_str().to_lowercase();
        out.push(format!("{a}-{b}"));
    }
    out
}

pub fn pgn_format(headers: &std::collections::HashMap<String, String>) -> String {
    headers
        .get("Format")
        .map(|s| s.trim().to_uppercase())
        .unwrap_or_default()
}

/// 与 Python 一致：以 `[Event` 作为新对局起点。
pub fn read_pgn_games(raw: &str) -> Vec<ParsedGame> {
    let raw = raw.replace("\r\n", "\n").replace('\r', "\n");
    let event_starts: Vec<usize> = event_line_re()
        .find_iter(&raw)
        .map(|m| m.start())
        .collect();
    if event_starts.is_empty() {
        return Vec::new();
    }
    let mut games = Vec::new();
    for i in 0..event_starts.len() {
        let start = event_starts[i];
        let end = event_starts
            .get(i + 1)
            .copied()
            .unwrap_or(raw.len());
        let chunk = raw[start..end].trim();
        if let Some(g) = parse_one_game_chunk(chunk) {
            games.push(g);
        }
    }
    games
}

fn parse_one_game_chunk(chunk: &str) -> Option<ParsedGame> {
    let lines: Vec<&str> = chunk.lines().collect();
    let mut headers = std::collections::HashMap::new();
    let mut movetext_lines: Vec<String> = Vec::new();
    let mut phase = "headers";
    let hr = header_re();
    for line in lines {
        let s = line.trim();
        if phase == "headers" {
            if s.is_empty() {
                phase = "moves";
                continue;
            }
            if s.starts_with('[') {
                if let Some(caps) = hr.captures(s) {
                    let key = caps.get(1)?.as_str().to_string();
                    let mut val = caps.get(2)?.as_str().replace("\\\"", "\"");
                    val = val.replace("\\\\", "\\");
                    headers.insert(key, val);
                }
                continue;
            }
            phase = "moves";
        }
        if phase == "moves" && !s.is_empty() {
            movetext_lines.push(s.to_string());
        }
    }
    if headers.is_empty() {
        return None;
    }
    Some(ParsedGame {
        headers,
        movetext_raw: movetext_lines.join(" "),
    })
}
